src/expression.rs

```diff
@@ -513,18 +513,64 @@ impl fmt::Display for Expression {
                 left,
                 operator,
                 right,
-            } => write!(f, "({}{}{})", left, operator, right),
+            } => {
+                format_inner_expression(f, left)?;
+                write!(f, "{}", operator)?;
+                format_inner_expression(f, right)
+            }
             Number(value) => write!(f, "{}", format_complex(value)),
             PiConstant => write!(f, "pi"),
             Prefix {
                 operator,
                 expression,
-            } => write!(f, "({}{})", operator, expression),
+            } => {
+                write!(f, "{}", operator)?;
+                format_inner_expression(f, expression)
+            }
             Variable(identifier) => write!(f, "%{}", identifier),
         }
     }
 }
 
+/// Utility function to wrap infix expressions that are part of an expression in parentheses, so
+/// that correct precedence rules are enforced.
+fn format_inner_expression(f: &mut fmt::Formatter, expression: &Expression) -> fmt::Result {
+    match expression {
+        Expression::Infix {
+            left,
+            operator,
+            right,
+        } => write!(f, "({left}{operator}{right})"),
+        _ => write!(f, "{expression}"),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        expression::{Expression, InfixOperator, PrefixOperator},
+        real,
+    };
+
+    #[test]
+    fn formats_nested_expression() {
+        let expression = Expression::Infix {
+            left: Box::new(Expression::Prefix {
+                operator: PrefixOperator::Minus,
+                expression: Box::new(Expression::Number(real!(3f64))),
+            }),
+            operator: InfixOperator::Star,
+            right: Box::new(Expression::Infix {
+                left: Box::new(Expression::PiConstant),
+                operator: InfixOperator::Slash,
+                right: Box::new(Expression::Number(real!(2f64))),
+            }),
+        };
+
+        assert_eq!(expression.to_string(), "-3*(pi/2)");
+    }
+}
```

Comment:
This is somewhat tangential to the parser fix, but I figured I'd take a stab at simplifying expression formatting here. Now, only nested infix expression will have surrounding parentheses applied. It is a lot less aggressive, and I think makes the resulting Quil easier to read. It probably would have made issues like this easier to spot.

A few more examples:

Before:
```
RZ((-pi)) 0
RZ(((-pi)/2)) 0
```

After:
```
RZ(-pi) 0
RZ(-pi/2) 0
```

Reply:
I get the feeling this may not work well with a "minus negative" expression like `2 - (-3)`. Do we have a set of tests that verify the written expression is parsed as itself, and is a "minus negative" one of the test cases?

Reply:
Functionally it is fine.

Something like:
```
-3 - -2
```

will get formatted:
```
-3--2
```

Both parse and run correctly (at least on the QVM). It isn't the prettiest, but I'd rather have this edge case then all the extra parentheses we had before. As a compromise, we could wrap nested prefix expressions in parentheses too.

Reply:
To start, I added test cases to the expression parser for the "minus negative" case.